File: questao_07/questao_07.go
package main

import (
	"fmt"
)

type Professor struct {
	nome    string
	escolas []*Escola
}

func (p *Professor) AdicionarEscola(escola *Escola) {
	p.escolas = append(p.escolas, escola)
	escola.AdicionarProfessor(p)
}

func (p Professor) ListarEscolas() {
	fmt.Printf("Professor %s leciona nas escolas: ", p.nome)
	for _, escola := range p.escolas {
		fmt.Print(escola.nome + " ")
	}
	fmt.Println()
}

type Escola struct {
	nome       string
	professores []*Professor
}

func (e *Escola) AdicionarProfessor(professor *Professor) {
	e.professores = append(e.professores, professor)
}

func (e Escola) ListarProfessores() {
	fmt.Printf("Escola %s tem os professores: ", e.nome)
	for _, professor := range e.professores {
		fmt.Print(professor.nome + " ")
	}
	fmt.Println()
}

func main() {
	escola1 := &Escola{nome: "Escola A"}
	escola2 := &Escola{nome: "Escola B"}

	professor1 := &Professor{nome: "Viviano"}
	professor2 := &Professor{nome: "Thiago"}

	professor1.AdicionarEscola(escola1)
	professor1.AdicionarEscola(escola2)
	professor2.AdicionarEscola(escola1)

	professor1.ListarEscolas()
	professor2.ListarEscolas()

	escola1.ListarProfessores()
	escola2.ListarProfessores()
}

File: questao_07/questao_07.java
import java.util.ArrayList;
import java.util.List;

class Professor {
    private String nome;
    private List<Escola> escolas;

    public Professor(String nome) {
        this.nome = nome;
        this.escolas = new ArrayList<>();
    }

    public String getNome() {
        return nome;
    }

    public void adicionarEscola(Escola escola) {
        escolas.add(escola);
        escola.adicionarProfessor(this);
    }

    public void listarEscolas() {
        System.out.printf("Professor %s leciona nas escolas: ", nome);
        for (Escola escola : escolas) {
            System.out.print(escola.getNome() + " ");
        }
        System.out.println();
    }
}

class Escola {
    private String nome;
    private List<Professor> professores;

    public Escola(String nome) {
        this.nome = nome;
        this.professores = new ArrayList<>();
    }

    public String getNome() {
        return nome;
    }

    public void adicionarProfessor(Professor professor) {
        professores.add(professor);
    }

    public void listarProfessores() {
        System.out.printf("Escola %s tem os professores: ", nome);
        for (Professor professor : professores) {
            System.out.print(professor.getNome() + " ");
        }
        System.out.println();
    }
}

public class questao_07 {
    public static void main(String[] args) {
        Escola escola1 = new Escola("Escola A");
        Escola escola2 = new Escola("Escola B");

        Professor professor1 = new Professor("Rodrigo");
        Professor professor2 = new Professor("Luan");

        professor1.adicionarEscola(escola1);
        professor1.adicionarEscola(escola2);
        professor2.adicionarEscola(escola1);

        professor1.listarEscolas();
        professor2.listarEscolas();

        escola1.listarProfessores();
        escola2.listarProfessores();
    }
}
